src/net/ifaddrs/mod.rs

```diff
@@ -0,0 +1,250 @@
+    BroadcastAddr(Option<IpAddr>),
+    DestinationAddr(Option<IpAddr>),
+    Neither,
+}
+
+
+impl Iterator for InterfaceAddrs {
+    type Item = InterfaceAddr;
+    fn next(&mut self) -> Option<InterfaceAddr> {
+        // If the current ifaddrs is None, there are no more ifaddrs to inspect
+        if self.current.is_none() {
+            return None;
+        }
+
+        // Workaround for the borrow checker being overzealous
+        // (without ptr_temp, p would technically "still be in use" when the
+        // loop ends, meaning we couldn't advance to the next struct)
+        let ptr_temp = self.current.clone();
+        let p = ptr_temp.as_ref().unwrap();
+
+        // Get a pointer to the interface's name
+        let name_ptr = p.ifa_name;
+        // Check that name_ptr isn't null.
+        if name_ptr.is_null() {
+            panic!("getifaddrs() gave an ifaddrs struct with a null ifa_name");
+        }
+        
+        // UNSAFETY: Constructing CStr from pointer. If this pointer is
+        // null it's a libc bug; it's checked above.
+        let name = unsafe { CStr::from_ptr(name_ptr)
+            .to_string_lossy()
+            .into_owned()};
+
+        // Interpret the flags field into a typed version of those flags
+        let flags = IffFlags::from_bits_truncate(p.ifa_flags);
+    
+        // Get std::net::IpAddr representations of the address and netmask
+        // UNSAFETY: sockaddr_to_ipaddr requires valid pointer.
+        let address = unsafe { sockaddr_to_ipaddr(p.ifa_addr) };
+        // UNSAFETY: sockaddr_to_ipaddr requires valid pointer.
+        let netmask = unsafe { sockaddr_to_ipaddr(p.ifa_netmask) };
+
+        // Figure out which ifu type is needed and create it
+        let ifu = if flags.contains(iff_flags::IFF_POINTOPOINT) {
+            // Point to point destination address
+            // UNSAFETY: sockaddr_to_ipaddr requires valid pointer.
+            let ifu_addr = unsafe { sockaddr_to_ipaddr(p.ifa_ifu) };
+            InterfaceIfu::DestinationAddr(ifu_addr)
+        } else if flags.contains(iff_flags::IFF_BROADCAST) {
+            // Broadcast address
+            // UNSAFETY: sockaddr_to_ipaddr requires valid pointer.
+            let ifu_addr = unsafe { sockaddr_to_ipaddr(p.ifa_ifu) };
+            InterfaceIfu::BroadcastAddr(ifu_addr)
+        } else { InterfaceIfu::Neither };
+
+        // Move along the list to the next ifaddrs struct
+        // UNSAFETY: *mut -> Option<&'static mut>. 
+        // This is known to be in valid memory or null.
+        self.current = unsafe { p.ifa_next.as_ref() };
+    
+        Some(InterfaceAddr {
+            name: name,
+            address: address,
+            netmask: netmask,
+            ifu: ifu,
+            flags: flags,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn tests_get_if_addrs() {
+        let ifs = super::InterfaceAddrs::query_system().unwrap();
+        for i in ifs {
+            println!("{}:", i.name);
```

Comment:
This isn't much of a test, because it doesn't verify anything.  It's really an example, and should go in the examples directory.  Making a proper test would be hard.  You'd probably have to compare against the output of `ifconfig`.